README.rst
===========
GitlabIRCed
===========


.. image:: https://img.shields.io/pypi/v/gitlabirced.svg
        :target: https://pypi.python.org/pypi/gitlabirced

.. image:: https://gitlab.com/palvarez89/gitlabirced/badges/master/pipeline.svg
        :target: https://gitlab.com/palvarez89/gitlabirced/commits/master

.. image:: https://gitlab.com/palvarez89/gitlabirced/badges/master/coverage.svg?job=tests-py36
        :target: https://gitlab.com/palvarez89/gitlabirced/commits/master

.. image:: https://readthedocs.org/projects/gitlabirced/badge/?version=latest
        :target: https://gitlabirced.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


IRC bot that connects with your GitLab projects


* Free software: MIT license
* Documentation: https://gitlabirced.readthedocs.io.


Features
--------

**GitlabIRCed** bot will let you connect your Gitlab projects with their IRC channels. There are main functions of this bot:

* **Hooks**: The bot will receive web hooks sent by Gitlab on different events.
  You can configure the bot to stream these events into one or many IRC channels
  and networks.
* **Watchers**: The bot can be configured to link an IRC channel to a project. The bot
  will watch the messages and give extra information when a MR or an Issue is mentioned.

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.5.0 (2019-07-24)
------------------

* Added support for SSL connections

0.4.2 (2019-07-17)
------------------

* Small release to fix the coverage badge in PyPI

0.4.1 (2019-07-17)
------------------

* Added support for Python 3.5. Now this version
  is also tested in the CI.
* Added support for Gitlab Personal Access Tocken in
  the 'watchers' section. Now you can watch private
  projects. Thanks Jonathan!
* Frozen dependencies in the requirements.txt file,
  to avoid random issues like the one fixed in fb99ad91
  (yaml.load vs yaml.safe_load).

0.4.0 (2018-11-11)
------------------

* Added support to ignore WIP merge requests.
* Added support for MR title changes.
* Improved reconnection on initial failure.
* Added support for MR revision changes.
* Improved anti-spam control. If the bot mentions an Issue/MR,
  it will not respond to a watcher command.

0.3.0 (2018-10-20)
------------------

* Added support for update MR assignee hooks.
* Improved reconnection failures.

0.2.3 (2018-10-18)
------------------

* Added support for update issue label hooks.
* Stopped ignoring repeated issue hooks.

0.2.2 (2018-09-06)
------------------

* Improved connection to channels that need registration to join.
* Fixed issue and mr detection on watchers.
* Include more information in logging lines.

0.2.1 (2018-09-03)
------------------

* Small bug fix release. We were missing some dependencies.

0.2.0 (2018-09-03)
------------------

* Added watchers support. Now the bot can be hanging
  on a channel giving useful information on MR an Issue
  mentions (e.g. !2, #59).
* Improved support of hook events.
* Improved reconnection of the bot.
* Improve logging, and introudce '-v' option to set the level
  of verbosity (e.g. -vvvv).
* Added authentication options (sasl and NickServ).
* Allow configuring the port used via the configuration file.
* Created plenty of scenario tests using behave.

0.1.3 (2018-08-23)
------------------

* Support added for 'merge_request'.

0.1.2 (2018-08-22)
------------------

* First release on PyPI.
* Support added for 'push' and 'issue' events.


tests/fake_helpers.py
class FakeConnection():
    def __init__(self):
        self.privmsgs = {}
        self.channels = []

    def privmsg(self, target, msg):
        msgs = self.privmsgs.get(target, [])
        msgs.append("(%s) %s" % (target, msg))
        self.privmsgs[target] = msgs

    def join(self, channel):
        self.channels.append(channel)

    def set_keepalive(self, interval):
        pass


class FakeBot():
    def __init__(self):
        self.connection = FakeConnection()


tests/test_irc_client.py
import unittest
from irc.client import Event

from fake_helpers import FakeConnection
from gitlabirced.irc_client import MyIRCClient


class BaseIRCClientTestCase(unittest.TestCase):
    def _fake_info(self, url, gitlabToken):
        return self.code, {'title': self.title,
                           'web_url': url}

    def setUp(self):
        self.connection = FakeConnection()
        self.target = '#target'

        watcher = {'network': 'freenode',
                   'channel': self.target,
                   'project': 'namespace/project'}
        self.mycli = MyIRCClient(
            [self.target], 'nick', 'freenode.org', 'freenode',
            watchers=[watcher], nickpass="p4ssw0rd")

        self.mycli._fetch_gitlab_info = self._fake_info

    def test_on_pubmsg_irrelevant(self):
        event = Event('pubmsg', '@somebody', self.target, ['a message'])

        # Receive a pubmsg
        self.mycli.on_pubmsg(self.connection, event)
        self.assertEqual(self.mycli.count_per_channel[self.target], 1)

        # Receive a second pubmsg
        self.mycli.on_pubmsg(self.connection, event)
        self.assertEqual(self.mycli.count_per_channel[self.target], 2)

        self.assertEqual(self.connection.privmsgs, {})

    def test_on_pubmsg_issue(self):
        self.title = "Title of the issue"
        self.code = 200

        expected_msg = ("(#target) Issue #12: Title of the issue https://"
                        "gitlab.com/api/v4/projects/namespace%2Fproject/"
                        "issues/12")

        self._test_on_pubmsg_generic('show me #12 plz', expected_msg)

    def test_on_pubmsg_mr(self):
        self.title = "Title of the mr"
        self.code = 200

        expected_msg = ("(#target) MR !12: Title of the mr https://"
                        "gitlab.com/api/v4/projects/namespace%2Fproject/"
                        "merge_requests/12")

        self._test_on_pubmsg_generic('show me !12 plz', expected_msg)

    def _test_on_pubmsg_generic(self, msg, expected):
        event = Event('pubmsg', '@somebody', self.target, [msg])

        # Receive a message with an issue number tag
        self.mycli.on_pubmsg(self.connection, event)
        # Expect an answer from the bot
        self.assertEqual(self.mycli.count_per_channel[self.target], 1)
        self.assertEqual(len(self.connection.privmsgs[self.target]), 1)
        self.assertEqual(
            self.connection.privmsgs[self.target][-1], expected)

        # Receive 'spam_threshold' more messages
        for i in range(self.mycli.spam_threshold):
            self.mycli.on_pubmsg(self.connection, event)
            self.assertEqual(self.mycli.count_per_channel[self.target], 2 + i)
        # Expect the bot to stay quiet
        self.assertEqual(len(self.connection.privmsgs[self.target]), 1)

        # Receive the same message once more
        self.mycli.on_pubmsg(self.connection, event)
        # Expect the bot to reply again
        self.assertEqual(self.mycli.count_per_channel[self.target], 17)
        self.assertEqual(len(self.connection.privmsgs[self.target]), 2)
        self.assertEqual(
            self.connection.privmsgs[self.target][-1], expected)

    def test_on_welcome(self):
        event = Event('welcome', 'freenode.net', 'me', ['WELCOME!'])

        # Expect an IDENTIFY message to NickServ
        self.mycli.on_welcome(self.connection, event)
        self.assertEqual(
            self.connection.privmsgs['NickServ'][-1],
            "(NickServ) IDENTIFY p4ssw0rd")
        # Expect some channels joined
        self.assertEqual(
            self.connection.channels,
            [self.target])


tests/test_gitlabirced.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `gitlabirced` package."""

import pytest

from click.testing import CliRunner

from gitlabirced import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface_no_args():
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 2
    assert 'Error: Missing argument "config-file"' in result.output


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    # result = runner.invoke(cli.main, ['tests/conf.yaml'])
    # assert result.exit_code == 0
    # assert 'gitlabirced.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help          Show this message and exit.' in help_result.output
    assert '-l, --log TEXT  Log output to this file' in help_result.output
    assert ('-v, --verbose   Verbose mode (-vvv for more, -vvvv max)'
            in help_result.output)


gitlabirced/cli.py
# -*- coding: utf-8 -*-

"""Console script for gitlabirced."""
import click
import copy
import logging
import signal
import sys
import threading
import yaml

from irc.client import is_channel

from .irc_client import connect_networks
from .http_server import MyHTTPServer, RequestHandler

cli_logger = logging.getLogger(__name__)


@click.command()
@click.argument('config-file', nargs=1)
@click.option('-v', '--verbose', count=True,
              help="Verbose mode (-vvv for more, -vvvv max)")
@click.option('-l', '--log', help="Log output to this file")
def main(config_file, verbose, log):
    _configure_logging(verbose, log)
    config = load_config(config_file)
    print(config)
    client = Client(config)
    client.start()
    signal.signal(signal.SIGINT, client.stop)


def load_config(config_file):
    try:
        with open(config_file, 'r') as stream:
            config = yaml.safe_load(stream)
            print("Configuration loaded %s" % config)
    except yaml.YAMLError as exc:
        print(exc)
        sys.exit(1)
    except IOError:
        print("File %s not found" % config_file)
        sys.exit(3)
    return config


class Client():
    def __init__(self, config):
        self.config = config
        self.all_bots = []

    def stop(self, sig=None, frame=None):
        print('You pressed Ctrl+C!')
        for b in self.all_bots:
            self.all_bots[b]['bot'].shutdown()
        self.httpd.shutdown()

    def start(self):
        """Console script for gitlabirced."""

        network_info = _get_channels_per_network(self.config)
        watchers = self.config.get('watchers')
        print('going to connect networks')
        self.all_bots = connect_networks(network_info, watchers)

        hooks = self.config.get('hooks', {})
        token = self.config['token']

        def run_server(addr, port):
            """Start a HTTPServer which waits for requests."""
            self.httpd = MyHTTPServer(token, hooks, self.all_bots,
                                      (addr, port), RequestHandler)
            thread = threading.Thread(target=self.httpd.serve_forever)
            thread.start()
            _, self.port_used = self.httpd.socket.getsockname()

        print('going to execute server')
        port = self.config.get('port', 0)
        if port == 0:
            cli_logger.warning('WARNING: Port not specified in the '
                               'configuration. A random one will be used')
        run_server('0.0.0.0', port)
        if port == 0:
            cli_logger.warning('WARNING: Port used {port}'
                               .format(port=self.port_used))
        print('Press Ctrl+C')

        return 0


def _get_channels_per_network(cfg):
    hooks = cfg.get('hooks', {})
    network_info = copy.deepcopy(cfg['networks'])
    for net_key in network_info:
        network_info[net_key]['channels'] = []

    for hook in hooks:
        network = hook['network']
        if network not in network_info:
            raise Exception("Network '{network}' not configured"
                            .format(network=network))

        reports = hook['reports']
        for ch in reports:
            current_channels = network_info[network]['channels']
            if is_channel(ch) and ch not in current_channels:
                print("Appending {channel} ({network})"
                      .format(channel=ch, network=network))
                network_info[network]['channels'].append(ch)

    watchers = cfg.get('watchers', {})
    print(watchers)
    for watcher in watchers:
        network = watcher['network']
        channel = watcher['channel']
        current_channels = network_info[network]['channels']
        if channel not in current_channels:
            print("Appending {channel} ({network})"
                  .format(channel=channel, network=network))
            network_info[network]['channels'].append(channel)

    return network_info


def _configure_logging(verbosity, output_file=None):
    """ Configures logging level in different ways.

    :param verbosity: The verbosity level (0-4)
      0: logging.WARNING
      1: logging.INFO
      2: logging.DEBUG
      3: (root) logging.INFO
      4: (root) logging.DEBUG
    :param output_file: If set, file to put the logs.
    """
    our_module_name = __name__.split('.')[0]
    our_logger = logging.getLogger(our_module_name)
    root_logger = logging.getLogger()

    our_level = None
    root_level = None

    if verbosity == 1:
        our_level = logging.INFO
    elif verbosity == 2:
        our_level = logging.DEBUG
    if verbosity == 3:
        root_level = logging.INFO
    if verbosity >= 4:
        root_level = logging.DEBUG

    if root_level:
        root_logger.setLevel(root_level)
    elif our_level:
        our_logger.setLevel(our_level)

    if output_file:
        handler = logging.FileHandler(output_file)
    else:
        handler = logging.StreamHandler()
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    handler.setFormatter(formatter)
    root_logger.addHandler(handler)


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


gitlabirced/irc_client.py
import irc.strings
import irc.client
import irc.bot

import logging
import threading
import time
import re
import requests
import ssl
import urllib

irc_client_logger = logging.getLogger(__name__)


class MyIRCClient(irc.bot.SingleServerIRCBot):
    def __init__(self, channels, nickname, server, net_name, port=6667,
                 watchers=None, nickpass=None, auth_type=None, ssl_conn=False):

        saslpass = nickpass
        if auth_type and auth_type.lower() != 'sasl':
            saslpass = None
        spec = irc.bot.ServerSpec(server, port, saslpass)

        if ssl_conn:
            ssl_factory = irc.connection.Factory(wrapper=ssl.wrap_socket)
            irc.bot.SingleServerIRCBot.__init__(
                self, [spec], nickname, nickname, connect_factory=ssl_factory)
        else:
            irc.bot.SingleServerIRCBot.__init__(
                self, [spec], nickname, nickname)

        # Monkey patch process_forever with a custom loop that we can stop
        def process_forever(timeout=0.2):
            while True:
                if self._stop_process_forever:
                    return
                self.reactor.process_once(timeout)
        self.reactor.process_forever = process_forever

        self._stop_process_forever = False
        self.channels_to_join = channels
        self.nickname = nickname
        self.nickpass = nickpass
        self.auth_type = auth_type
        self.server = server
        self.net_name = net_name
        self.port = port
        self.watchers = watchers
        self.last_mention = {}
        self.count_per_channel = {}
        self.spam_threshold = 15
        self.key_template = '{kind}{channel}{number}'
        self.ping_configured = False

    def shutdown(self):
        self._stop_process_forever = True
        self.reactor.disconnect_all()

    def _log_warning(self, msg):
        irc_client_logger.warning("(%s) %s" % (self.net_name, msg))

    def _log_info(self, msg):
        irc_client_logger.info("(%s) %s" % (self.net_name, msg))

    def _log_debug(self, msg):
        irc_client_logger.debug("(%s) %s" % (self.net_name, msg))

    def _log_error(self, msg):
        irc_client_logger.error("(%s) %s" % (self.net_name, msg))

    def on_privnotice(self, c, e):
        self._log_info("PRIVNOTICE: %s" % e.arguments[0])

    def on_nicknameinuse(self, c, e):
        # TODO: Try to retake nick? Or use a different one?
        # Probably best to try to retake nick, see gerrit bot
        # for implementation reference.
        pass

    def on_error(self, c, e):
        error = ""
        if len(e.arguments) > 0:
            error = e.arguments[0]
        else:
            error = str(e)
        self._log_error("ERROR: %s" % error)

    def on_welcome(self, connection, event):
        if not self.ping_configured:
            connection.set_keepalive(10)
            self.ping_configured = True

        if self.nickpass:
            if (not self.auth_type) or self.auth_type.lower() == 'nickserv':
                connection.privmsg('NickServ', 'IDENTIFY {password}'.format(
                    password=self.nickpass))

        self._join_channels(connection)

    def _join_channels(self, connection):
        for ch in self.channels_to_join:
            connection.join(ch)

    def on_nochanmodes(self, connection, event):
        # We couldn't join a channel, wait and retry
        timeout = 10
        self._log_info("NOCHANMODES retry to join in %s seconds" % timeout)
        time.sleep(timeout)
        self._join_channels(connection)

    def on_disconnect(self, connection, event):
        if not self._stop_process_forever:
            timeout = 10
            self._log_info("DISCONNECT reconnecting in %s seconds" % timeout)
            connected = False
            while not connected:
                # Wait a bit to avoid throttling
                time.sleep(timeout)
                try:
                    connection.reconnect()
                except irc.client.ServerConnectionError as e:
                    self._log_error(
                        "Failed to reconnect. Reconnecting in %s seconds."
                        "Reason: %s" % (timeout, e))
                else:
                    connected = True

    def _update_count(self, channel):
        count = self.count_per_channel.get(channel, 0) + 1
        self.count_per_channel[channel] = count

    def on_pubmsg(self, c, e):
        on_channel = e.target
        self._log_info("PUBMSG on channel %s: %s" % (
            on_channel, e.arguments[0]))

        if not self.watchers:
            return

        self._update_count(on_channel)

        for w in self.watchers:
            if not (w['network'] == self.net_name and
                    w['channel'] == on_channel):
                continue

            msg = e.arguments[0].split()
            mr_regex = r'!([0-9]+)'
            issue_regex = r'#([0-9]+)'
            for m in msg:
                issue_match = re.match(issue_regex, m)
                if issue_match:
                    self._log_info("PUBMSG contains ISSUE mention (%s)" % m)
                    self._fetch_and_say(c, 'issue', issue_match.group(1), w)

                mr_match = re.match(mr_regex, m)
                if mr_match:
                    self._log_info("PUBMSG contains MR mention (%s)" % m)
                    self._fetch_and_say(
                        c, 'merge_request', mr_match.group(1), w)

            # Only one watcher allowed per channel. Stop.
            break

    def _mentioned_recently(self, channel, kind, number):
        key = self.key_template.format(
                kind=kind, channel=channel, number=number)
        last_time = self.last_mention.get(key)
        if last_time is None:
            return False
        if ((self.count_per_channel.get(channel, 0)-last_time) <=
                self.spam_threshold):
            return True
        return False

    def _update_mentions(self, channel, kind, number):
        key = self.key_template.format(
                kind=kind, channel=channel, number=number)
        self.last_mention[key] = self.count_per_channel.get(channel, 0)

    def _fetch_and_say(self, c, kind, number, watcher):
        server = watcher.get('server', 'https://gitlab.com')
        target = watcher['channel']
        project_encoded = urllib.parse.quote(watcher['project'], safe='')
        gitlabToken = watcher.get('gitlabtoken', '')

        if self._mentioned_recently(target, kind, number):
            self._log_debug("_fetch_and_say decides not to react")
            return

        if kind == 'issue':
            url_template = 'api/v4/projects/{project}/issues/{number}'
            prefix_template = 'Issue #{number}:'
        elif kind == 'merge_request':
            url_template = 'api/v4/projects/{project}/merge_requests/{number}'
            prefix_template = 'MR !{number}:'

        url = urllib.parse.urljoin(server, url_template.format(
            project=project_encoded, number=number))

        self._log_debug("_fetch_and_say will query %s" % url)

        status, info = self._fetch_gitlab_info(url, gitlabToken)
        if status != 200:
            self._log_error("Failed to query %s" % url)
            return

        self._log_info("Title %s" % info['title'])
        self._log_info("URL   %s" % info['web_url'])
        prefix = prefix_template.format(number=number)
        info_text = '{prefix} {title} {url}'.format(
                prefix=prefix, title=info['title'], url=info['web_url'])
        self._log_debug("_fetch_and_say sending to %s - %s" % (
            target, info_text))
        c.privmsg(target, info_text)
        self._update_mentions(target, kind, number)

    def _fetch_gitlab_info(self, url, gitlabToken):
        data = requests.get(
               url,
               headers={'PRIVATE-TOKEN': '{}'.format(gitlabToken)}
            )
        return data.status_code, data.json()


def connect_networks(networks, watchers):
    """ Connects to all the networks configured in the config file.

    Returns a dictionary using the same keys as in the config file
    containing process and bot object.
    """

    print("conf taken %s " % networks)
    result = {}
    for net in networks:
        server = networks[net]['url']
        port = networks[net]['port']
        nick = networks[net]['nick']
        channels = networks[net]['channels']
        password = networks[net].get('pass')
        auth_type = networks[net].get('auth')
        ssl_conn = networks[net].get('ssl', False)

        # TODO: Only use password if auth is set to 'sasl'
        bot = MyIRCClient(channels, nick, server, net, port=port,
                          watchers=watchers, nickpass=password,
                          auth_type=auth_type, ssl_conn=ssl_conn)
        irc_client_logger.info('Starting client for server %s' % server)
        thread = threading.Thread(target=bot.start)
        thread.start()
        # Schedule a reconnection
        # Submitted upstream: https://github.com/jaraco/irc/pull/156
        bot.recon.run(bot)

        result[net] = {
            'process': thread,
            'bot': bot
        }

    return result


gitlabirced/http_server.py
from http.server import HTTPServer
from http.server import BaseHTTPRequestHandler
import json
import logging

http_server_logger = logging.getLogger(__name__)


class MyHTTPServer(HTTPServer):

    def __init__(self, token, hooks, bots, *args, **kw):
        HTTPServer.__init__(self, *args, **kw)
        self.token = token
        self.hooks = hooks
        self.bots = bots


class RequestException(Exception):
    def __init__(self, code, status):
        self.code = code
        self.status = status
        http_server_logger.error('%s - %s' % (self.code, self.status))


class RequestHandler(BaseHTTPRequestHandler):
    """A POST request handler."""

    def _check_token(self):
        # get gitlab secret token
        gitlab_token_header = self.headers.get('X-Gitlab-Token')

        if not gitlab_token_header:
            raise RequestException(400, "'X-Gitlab-Token' header not found")

        # get token from config file
        gitlab_token = str(self.server.token)

        # Check if the gitlab token is valid
        if gitlab_token_header != gitlab_token:
            raise RequestException(401, "Gitlab token not authorized")

    def _check_and_get_request_data(self):
        # get payload
        header_length = int(self.headers.get('content-length', "0"))
        json_payload = self.rfile.read(header_length)

        if len(json_payload) == 0:
            raise RequestException(400, "Request didn't contain data")

        try:
            try:
                json_params = json.loads(json_payload)
            except TypeError:
                # Python 3.5 json.loads doesn't support binary input
                # https://docs.python.org/3/whatsnew/3.6.html#json
                json_params = json.loads(json_payload.decode('utf-8'))
        except json.decoder.JSONDecodeError:
            raise RequestException(400, "JSON data couldn't be parsed")

        object_kind = json_params.get('object_kind')
        if not object_kind:
            raise RequestException(400, "Missing 'object_kind'")

        if object_kind not in ['push', 'issue', 'merge_request']:
            raise RequestException(400, "object_kind '%s' not supported" %
                                   object_kind)
        return json_params

    def do_POST(self):
        http_server_logger.info("Hook received")

        json_params = None
        try:
            self._check_token()
            json_params = self._check_and_get_request_data()

            handler = getattr(
                self, '_handle_%s' % json_params.get('object_kind'))
            handler(json_params)
        except RequestException as re:
            if json_params:
                http_server_logger.debug('JSON PARAMS: %s' % json_params)
            http_server_logger.exception(re.status)
            self.send_response(re.code, re.status)
        except Exception:
            http_server_logger.exception("Internal server error")
            self.send_response(500, "Internal server error")
        finally:
            self.end_headers()

    def _handle_push(self, json_params):
        http_server_logger.info('handling push')

        try:
            project = json_params['project']['path_with_namespace']
            project_name = json_params['project']['name']
            user = json_params['user_username']
            commits = json_params['commits']
            num_commits = json_params.get('total_commits_count')
            branch_name = json_params['ref']
            ref_after = json_params['after']
            ref_prefix = 'refs/heads/'
            if branch_name.startswith(ref_prefix):
                branch_name = branch_name[len(ref_prefix):]
        except KeyError:
            raise RequestException(400, "Missing data in the request")

        if not num_commits:
            # Branch created or deleted
            action = 'created'
            if ref_after == '0000000000000000000000000000000000000000':
                action = 'deleted'
            msg = ('{user} {action} branch {branch_name} on {project_name}.'
                   .format(user=user, action=action,
                           branch_name=branch_name,
                           project_name=project_name))
        else:
            last_commit = commits[-1]
            last_commit_msg = last_commit['message'].split('\n')[0].strip()
            pre_msg = ('{user} pushed on {project_name}@{branch_name}:'
                       .format(user=user, project_name=project_name,
                               branch_name=branch_name))
            if num_commits == 1:
                msg = ('{pre_msg} {last_commit_msg}'
                       .format(pre_msg=pre_msg,
                               last_commit_msg=last_commit_msg))
            else:
                msg = ('{pre_msg} {num_commits} commits (last: '
                       '{last_commit_msg})'
                       .format(pre_msg=pre_msg,
                               num_commits=num_commits,
                               last_commit_msg=last_commit_msg))

        self._send_message_to_all('push', project, msg, branch=branch_name)

    def _handle_issue(self, json_params):
        http_server_logger.info('handling issue')

        try:
            user = json_params['user']['username']
            project_name = json_params['project']['name']
            project = json_params['project']['path_with_namespace']
            issue = json_params['object_attributes']
            issue_number = issue['iid']
            issue_title = issue['title'].strip()
            issue_action = issue['action']
            url = issue['url']
        except KeyError:
            raise RequestException(400, "Missing data in the request")

        display_action = self.simple_past(issue_action)
        hook_key = 'issue'
        if issue_action == 'update':
            changes = json_params['changes']
            if "labels" in changes:
                previous = [label['title']
                            for label in changes['labels']['previous']]
                current = [label['title']
                           for label in changes['labels']['current']]
                added = list(set(current).difference(previous))
                removed = list(set(previous).difference(current))
                # Be deterministic
                added.sort()
                removed.sort()
                added_quoted = ", ".join(["'%s'" % l for l in added])
                removed_quoted = ", ".join(["'%s'" % l for l in removed])
                chg_msg = ""
                if added:
                    chg_msg += "added %s " % added_quoted
                if added and removed:
                    chg_msg += "and "
                if removed:
                    chg_msg += "removed %s " % removed_quoted
                chg_msg += "label(s) to"
                hook_key = 'issue_label'
            else:
                # Unsupported update hook
                self.send_response(200, "OK")
                return
            display_action = chg_msg

        msg = ('{user} {display_action} issue #{issue_number} '
               '({issue_title}) on {project_name} {url}'
               .format(user=user, display_action=display_action,
                       issue_number=issue_number, issue_title=issue_title,
                       project_name=project_name, url=url))

        self._send_message_to_all(
            hook_key, project, msg, 'issue', issue_number)

    def _handle_merge_request(self, json_params):
        http_server_logger.info('handling merge_request')

        try:
            user = json_params['user']['username']
            request = json_params['object_attributes']
            project = request['target']['path_with_namespace']
            project_name = request['target']['name']
            from_branch = request['source_branch']
            to_branch = request['target_branch']

            request_number = request['iid']
            request_title = request['title'].strip()
            request_action = request['action']
            request_wip = request['work_in_progress']
            url = request['url']
        except KeyError:
            raise RequestException(400, "Missing data in the request")

        display_action = self.simple_past(request_action)
        hook_key = 'merge_request'
        if request_action == 'update':
            changes = json_params['changes']
            chg_msg = ""
            if "oldrev" in request:
                hook_key = 'merge_request_update'
                chg_msg = "updated commits on"

            elif "title" in changes:
                def is_title_wip(title):
                    title_lower = title.lower()
                    flags = ["[WIP]", "WIP:", "WIP "]
                    for flag in flags:
                        if flag.lower() in title_lower:
                            return True
                    return False
                was_wip = is_title_wip(changes['title']['previous'])
                if was_wip and not request_wip:
                    chg_msg = "opened (was WIP)"
                else:
                    hook_key = 'merge_request_title'
                    chg_msg = "changed title of"

            elif "assignee" in changes:
                hook_key = 'merge_request_assignee'
                previous = changes['assignee']['previous']
                current = changes['assignee']['current']
                # If current is empty, it's an unassignment
                if current:
                    chg_msg = "assigned %s to" % current['username']
                else:
                    chg_msg = "unassigned %s from" % previous['username']
            else:
                # Unsupported update hook
                self.send_response(200, "OK")
                return
            display_action = chg_msg

        msg = ('{user} {display_action} MR !{request_number} '
               '({from_branch}->{to_branch}: {request_title}) '
               'on {project_name} {url}'
               .format(user=user, display_action=display_action,
                       request_number=request_number, from_branch=from_branch,
                       to_branch=to_branch, request_title=request_title,
                       project_name=project_name, url=url))

        self._send_message_to_all(
            hook_key, project, msg, 'merge_request', request_number,
            is_wip=request_wip)

    @staticmethod
    def simple_past(verb):
        if verb.endswith('ed'):
            return verb
        if not verb.endswith('e'):
            verb = verb + 'e'
        verb = verb + 'd'
        return verb

    def _send_message_to_all(self, hook_key, project, msg,
                             kind=None, number=None, branch=None,
                             is_wip=False):
        hooks = self.server.hooks
        bots = self.server.bots
        for h in hooks:
            if h['project'] == project:
                network = h['network']
                reports = h['reports']
                branches = h.get('branches', '').split()
                show_wips = h.get('wip', False)
                bot = bots[network]['bot']
                if branch and branch not in branches:
                    continue
                if not show_wips and is_wip:
                    continue
                for r in reports:
                    if hook_key in reports[r]:
                        http_server_logger.info('sending to %s, in network %s'
                                                % (r, network))
                        bot.connection.privmsg(r, msg)
                        if kind and number:
                            bot._update_mentions(r, kind, number)

        self.send_response(200, "OK")
